lib/gc/Transforms/CPUPhysicalRegisterPass.cpp

```diff
@@ -0,0 +1,2978 @@
+
+    //  1.b. What operands are needed to find in the current group, and where
+    //  can they be obtained ?
+
+    //  Thanks to 1.a, we get the result generated by the operations of
+    //  each group, and this result will use `scf.yield` to generate a
+    //  new result. Since the scope of the parent block of mlir is covered
+    //  the current operation, the current operation does not need to pass
+    //  these `for loop result` to the `iterArgs` of the required `for loop`.
+    //  It only needs to replace the operand of the current operation with the
+    //  corresponding `for loop yield result`.
+
+    // However, for some operations that are not DPS, we need to canonicalize
+    // them. Canonicalization means that the operand of this operation is a
+    // vector but we can't get this vector due to it locates in another block
+    // which has a different scope. Therefore, it is necessary to write the
+    // vector results into a temporary tensor to save it. Then the vector
+    // needs to be read from the tensor before the current operation operate
+    // on it. Therefore,  `empty tensor`, `transfer_write` and `transfer_read`
+    // need to be inserted at target place.
+    if (enableDebugPrinter) {
+      printGroupOps(fusion.getGroupOperationFusion().getOpGroups());
+      LDBG("___________ before analysis ________________");
+    }
+    fusion.canonicalizeEachOperationGroup();
+    if (enableDebugPrinter) {
+      LDBG("___________ after analysis ________________");
+      printGroupOps(fusion.getGroupOperationFusion().getOpGroups());
+    }
+
+    loopGenerator.setVectorBaseFusion(fusion.getGroupOperationFusion());
+    // Speical Operation Canonicalization
+    loopGenerator.canonicalizeSpecialOperation();
+
+    // 2.Generate vectorized IR for each operation group
+    for (size_t idx = 0; idx < fusionStrategy.getOpGroups().size(); ++idx)
+      loopGenerator.generateGroupOpVectorizedIR(idx);
+
+    // 3. Some IR cleanup work
+    DominanceInfo domInfo;
+    eliminateCommonSubExpressions(
+        rewriter, domInfo, loopGenerator.getVectorBasedFusion().getFunction());
+  } else {
+    // TODO: need to add directly canonicalize operations logic
+    // generateGroupOpVectorizedIR(idx, grp, fusionStrategy.opGroupIndexMap);
+  }
+}
+
+///
+void ForLoopGenerator::setOperationCorrectOperand(
+    Operation *op, const DenseMap<Operation *, AffineMap> &opPermuationMap,
+    GenerateLoopHelper &loopHelperParam) {
+  for (auto [idx, opd] : llvm::enumerate(op->getOperands())) {
+    if (not loopHelperParam.originalOperandLoopArgsMap.contains(opd))
+      continue;
+
+    Value loopArg = loopHelperParam.originalOperandLoopArgsMap[opd];
+    if (not loopHelperParam.currentLoopStateIdxMap.contains(loopArg))
+      continue;
+
+    op->setOperand(
+        idx,
+        loopHelperParam
+            .loopIterArgs[loopHelperParam.currentLoopStateIdxMap.at(loopArg)]);
+  }
+  int offset = isa<vector::TransferWriteOp>(op) ? 2 : 1;
```

Comment:
what is the meaning of 2 and 1 here?

Reply:
The offset index of the operand. Will rename it as `operandOffset`.